internal/cmd/gocdk/fswatch.go

```diff
@@ -0,0 +1,205 @@
+		dst  **dirSnapshot
+		path string
+	}
+
+	// Depth-first search over the directory structure.
+	var root *dirSnapshot
+	stack := []walkNode{{&root, path}}
+	for len(stack) > 0 {
+		// Pop next node off stack.
+		curr := stack[len(stack)-1]
+		stack = stack[:len(stack)-1]
+
+		// Read directory and create snapshot.
+		contents, err := ioutil.ReadDir(curr.path)
+		if err != nil {
+			return nil, xerrors.Errorf("snapshot tree %s: %w", path, err)
+		}
+		snap := newDirSnapshot(contents)
+		*curr.dst = snap
+
+		// Push subdirectories onto stack in reverse order, so they are popped in
+		// ascending order.
+		for i := len(snap.entries) - 1; i >= 0; i-- {
+			ent := snap.entries[i]
+			if !ent.mode.IsDir() {
+				continue
+			}
+			stack = append(stack, walkNode{&ent.snapshot, filepath.Join(curr.path, ent.name)})
+		}
+	}
+	return root, nil
+}
+
+// sameAs compares the directory snapshot to the given filesystem directory.
+func (snap *dirSnapshot) sameAs(path string) (bool, error) {
+	// TODO(light): This doesn't require a full walk in the case where the
+	// directory changed.
+	currSnap, err := snapshotTree(path)
+	if err != nil {
+		return false, xerrors.Errorf("compare directory %s to snapshot: %w", path, err)
+	}
+	return snap.recursiveEqual(currSnap), nil
+}
+
+// equal compares two directory snapshots for equality non-recursively.
+func (snap *dirSnapshot) equal(snap2 *dirSnapshot) bool {
+	if snap == nil && snap2 == nil {
+		return true
+	}
+	if snap == nil || snap2 == nil {
+		return false
+	}
+	if len(snap.entries) != len(snap2.entries) {
+		return false
+	}
+	for i := range snap.entries {
+		if !snap.entries[i].equal(snap2.entries[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+// recursiveEqual compares two directory snapshots for equality recursively.
+func (snap *dirSnapshot) recursiveEqual(snap2 *dirSnapshot) bool {
+	type walkNode struct {
+		snap1, snap2 *dirSnapshot
+	}
+
+	// Depth-first traversal.
+	stack := []walkNode{{snap, snap2}}
+	for len(stack) > 0 {
+		curr := stack[len(stack)-1]
+		stack = stack[:len(stack)-1]
+		if !curr.snap1.equal(curr.snap2) {
+			return false
+		}
+		for i := range curr.snap1.entries {
+			if curr.snap1.entries[i].snapshot == nil {
+				continue
+			}
+			stack = append(stack, walkNode{
+				curr.snap1.entries[i].snapshot,
+				curr.snap2.entries[i].snapshot,
+			})
+		}
+	}
+	return true
+}
+
+type dirSnapshotEntry struct {
+	name    string
+	mode    os.FileMode
+	modTime time.Time
+
+	// snapshot is filled in by snapshotTree if the entry is a directory.
+	snapshot *dirSnapshot
+}
+
+// equal compares two entries for equality non-recursively.
+func (ent *dirSnapshotEntry) equal(ent2 *dirSnapshotEntry) bool {
+	if ent == nil && ent2 == nil {
+		return true
+	}
+	if ent == nil || ent2 == nil {
+		return false
+	}
+	return ent.name == ent2.name && ent.mode == ent2.mode && ent.modTime.Equal(ent2.modTime)
+}
+
+// isIgnoredForWatch reports whether a given filename is ignored for the
```

Comment:
Do we care about anything other than `.go` files?

Reply:
Yes, because there could be template files or other resources that the binary depends on. (Also Terraform files would need to be included.)

Reply:
Right.

Reply:
I wonder if we should make this configurable somewhere... e.g., if the user's app writes some files, it could go wrong. Maybe add a TODO?

Reply:
To-Done.